rmw_connext_cpp/src/functions.cpp

```diff
@@ -164,62 +212,73 @@ rmw_create_publisher(
     rmw_set_error_string("callbacks handle is null");
     return NULL;
   }
-  std::string type_name = std::string(callbacks->package_name) + "::msg::dds_::" +
-    callbacks->message_name + "_";
+  std::string type_name = _create_type_name(callbacks, "msg");
+  // Past this point, a failure results in unrolling code in the goto fail block.
```

Comment:
A good discussion which mentions multiple pros / cons / alternatives: http://stackoverflow.com/questions/1073397/while1-break-instead-of-goto